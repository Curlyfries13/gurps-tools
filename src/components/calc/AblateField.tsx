import React, { useState } from 'react';
import { connect, ConnectedProps } from 'react-redux';

import { updateArmor } from '/src/redux/actions/armorActions';
import { RootState } from '/src/redux/reducers';

import { Armor } from '/src/types';
import { ensureFound } from '/src/utils/jsUtils';

const AblateField = ({ armorId, armor, updateArmor }: Props) => {
  // TODO: wire up vlaidation and errors
  const [errors, setErrors]: [any, Function] = useState({});
  const [warn, setWarn]: [any, Function] = useState({});
  const id: string = armorId + '-ablateBase';

  function handleValueChange(event: React.ChangeEvent<HTMLInputElement>): void {
    const { value } = event.target;
    const base: number = parseInt(value, 10) || 0;
    const updatedArmor: Armor = { ...armor, ablateBase: base };
    updateArmor(updatedArmor);
  }

  // TODO: wire up vlaidation
  function ablateIsValid(armor: Armor) {
    const errors: any = {};
    const warn: any = {};
    if (!(armor.ablateBase >= 0))
      errors.ablateBase = 'Ablate Base must be a positive value';
    if (!(armor.ablateBase == 0))
      warn.zero = 'Ablate Base should be a positive value';
    setErrors(errors);
    setWarn(warn);
    return Object.keys(errors).length === 0;
  }

  // TODO: update accessibility label to update with
  return (
    <div className='row'>
      <label htmlFor={id} className='col col-form-label'>
        Ablate Base
      </label>
      <div className='col'>
        <input
          aria-label="this armor's ablative base"
          className='form-control'
          id={id}
          value={armor.ablateBase}
          onChange={handleValueChange}
        />
      </div>
    </div>
  );
};

interface OwnProps {
  armorId: number;
}

function mapStateToProps(state: RootState, ownProps: OwnProps) {
  const armor: Armor = ensureFound<Armor>(
    state.armorStack.find((armor: Armor) => {
      return armor.id === ownProps.armorId;
    })
  );
  return {
    armor: armor,
  };
}

const mapDispatchToProps = {
  updateArmor,
};

const connector = connect(mapStateToProps, mapDispatchToProps);

type Props = ConnectedProps<typeof connector> & OwnProps;

export default connector(AblateField);
