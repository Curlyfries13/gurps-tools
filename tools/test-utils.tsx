// Using code from https://redux.js.org/recipes/writing-tests#connected-components
import React from 'react';

// test-utils.js
import { render } from '@testing-library/react';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
// Import your own reducer
//
// console.log(React.createElement);

import * as reducer from 'src/redux/reducers';


const customRender = (
  ui,
  {
    initialState,
    store = createStore(reducer, initialState),
    ...renderOptions
  } = {}
) => {
  const Wrapper =({ children }) => {
    return <Provider store={store}>{children}</Provider>;
  }
  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

// re-export everything
export * from '@testing-library/react';
// override render method
export { customRender as render };
