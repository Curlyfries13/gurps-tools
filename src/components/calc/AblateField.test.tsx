import React from 'react';
import { RenderResult } from '@testing-library/react';
import { render } from 'tools/test-utils';

import AblateField from './AblateField';

const mockArmor = { dr: 0, id: 0 };
const mockState = { armorStack: [mockArmor] };

it('matches snapshot', async () => {
  const component: RenderResult = render(<AblateField armorId={0} />, {
    initialState: mockState,
  });
  expect(component).toMatchSnapshot();
});
