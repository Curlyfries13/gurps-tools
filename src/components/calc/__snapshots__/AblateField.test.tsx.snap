// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="row"
      >
        <label
          class="col col-form-label"
          for="0-ablateBase"
        >
          Ablate Base
        </label>
        <div
          class="col"
        >
          <input
            aria-label="this armor's ablative base"
            class="form-control"
            id="0-ablateBase"
            value=""
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="row"
    >
      <label
        class="col col-form-label"
        for="0-ablateBase"
      >
        Ablate Base
      </label>
      <div
        class="col"
      >
        <input
          aria-label="this armor's ablative base"
          class="form-control"
          id="0-ablateBase"
          value=""
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
